src/main/kotlin/rx/lang/kotlin/observables.kt

```diff
@@ -58,34 +58,20 @@ public fun <T, R> Observable<T>.lift(operator : (Subscriber<in R>) -> Subscriber
 /**
  * Returns [Observable] that requires all objects to be non null. Raising [NullPointerException] in case of null object
  */
-public fun <T : Any> Observable<T?>.requireNoNulls() : Observable<T> = lift { s ->
-    subscriber<T?>().
-            onCompleted { s.onCompleted() }.
-            onError { t -> s.onError(t) }.
-            onNext { v -> if (v == null) throw NullPointerException("null element found in rx observable") else s.onNext(v) }
-}
+public fun <T : Any> Observable<T?>.requireNoNulls() : Observable<T> = map { it ?: throw NullPointerException("null element found in rx observable") }
 
 /**
  * Returns [Observable] with non-null generic type T. Returned observable filter out null values
  */
-public fun <T : Any> Observable<T?>.filterNotNull() : Observable<T> = lift { s ->
-    subscriber<T?>().
-            onCompleted { s.onCompleted() }.
-            onError { t -> s.onError(t) }.
-            onNext { v -> if (v != null) s.onNext(v) }
-}
+public fun <T : Any> Observable<T?>.filterNotNull(): Observable<T> = filter { it != null }.map { it as T }
 
 /**
  * Returns Observable that wrap all values into [IndexedValue] and populates corresponding index value.
  * Works similar to [kotlin.withIndex]
  */
-public fun <T> Observable<T>.withIndex() : Observable<IndexedValue<T>> = lift { s ->
+public fun <T> Observable<T>.withIndex() : Observable<IndexedValue<T>> {
     var index = 0
```

Comment:
Not. E.g.,
```
o2 = o1.withIndex()
o2.subscribe(s1)
o2.subscribe(s2)
```
Each `subscribe` should use a new `index`.

I know nothing about Kotlin. But in RxJava, you can implement it using `o.zipWith(Observable.range(0, Integer.MAX_VALUE))`.